deployment/src/main.rs

```diff
@@ -0,0 +1,489 @@
+}
+
+fn run_ssh_command<P1: AsRef<Path>>(host: &str, key_file: P1, args: &[&str]) -> Result<()> {
+    info!("Running `ssh {}", args.join(" "));
+
+    let result = Command::new("ssh")
+        .arg("-i")
+        .arg(key_file.as_ref())
+        .arg(format!("ubuntu@{}", host))
+        .args(args)
+        .status()?;
+
+    if !result.success() {
+        error!("Failed to run ssh.");
+        return Err(Error::msg("Failed run SSH command."));
+    }
+
+    Ok(())
+}
+
+async fn associate_public_ip(
+    ec2_client: &aws_sdk_ec2::Client,
+    instance_id: &str,
+    ip_addr: &str,
+) -> Result<()> {
+    info!("Associating elastic IP address with instance...");
+    ec2_client
+        .associate_address()
+        .instance_id(instance_id)
+        .allow_reassociation(true)
+        .public_ip(ip_addr)
+        .send()
+        .await?;
+
+    info!("Public IP {} now associated with {}.", ip_addr, instance_id);
+
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // You must set the access key ID associated with the IAM
+    // user that will run the deployment.
+    assert_env_set("AWS_ACCESS_KEY_ID");
+
+    // You must set the secret access key corresponding to
+    // AWS_ACCESSS_KEY_ID. This is the IAM user's secret.
+    assert_env_set("AWS_SECRET_ACCESS_KEY");
+
+    assert_env_set("EC2_SECRET_KEY");
+
+    assert_env_set("ELASTIC_IP_ADDRESS");
+
+    assert_env_set("DOCKER_REPOSITORY");
+
+    env_logger::init();
+
+    let public_ip_addr = std::env::var("ELASTIC_IP_ADDRESS").unwrap();
+    let docker_repository = std::env::var("DOCKER_REPOSITORY").unwrap();
+
+    let sdk_config = aws_config::from_env().region("us-west-2").load().await;
+
+    let ec2_client = aws_sdk_ec2::Client::new(&sdk_config);
+
+    let image = get_image(&ec2_client).await?;
+    info!("Using image {:#?}", image);
+
+    let instance = create_ec2_instance(&ec2_client, &image).await?;
+    let instance_id = instance.instance_id().unwrap();
+
+    let ip_name = get_ip_name(&ec2_client, &instance_id).await?;
+
+    info!("Instance DNS name {}", ip_name);
+
+    let key_file = install_ssh_key(&ip_name)?;
+
+    let artifacts_base_dir = PathBuf::from("./").canonicalize()?;
+
+    info!(
+        "Artifacts base directory = {}",
+        artifacts_base_dir.to_string_lossy()
+    );
+
+    copy_artifacts(
+        &ip_name,
+        &key_file,
+        artifacts_base_dir.join("./ui/frontend/build"),
+        "",
+    )?;
+    copy_artifacts(
+        &ip_name,
+        &key_file,
+        artifacts_base_dir.join("./target/release/ui"),
+        "ui",
+    )?;
+    copy_artifacts(&ip_name, &key_file, artifacts_base_dir.join("compiler/fetch.sh"), "fetch.sh")?;
+
+    run_ssh_command(
+        &ip_name,
+        &key_file,
+        &["./fetch.sh", &docker_repository],
+    )?;
+
+    run_ssh_command(
+        &ip_name,
+        &key_file,
+        &["sudo", "systemctl", "start", "playground"],
+    )?;
+
+    std::fs::remove_file(&key_file)?;
```

Comment:
Should add a rollback mechanism to terminate the new instance if setting up the instance fails after creating it, but before the the IP association.